Set expiration per session.

clems4ever opened the pull request:
In some cases it might be legitimate to override the session expiration. For
instance, in an authentication workflow, one might want to check the
"Remember me" option and in that case the session cookie should not expire
anymore while by default all new cookies should inherit from the default
configuration.

This change will allow a user to call SetExpiration(...) on a storer
so that the new expiration is applied to that session. This updates the cookie
with the new expiration. Please note that the regenerating a session will reset
the expiration to the default value which was globally configured.

clems4ever commented:
Hello @savsgio , I eventually moved the update method to the storer so that the user can
```
storer := session.Get(...)
storer.SetExpiration(3 * time.Hour)
session.Save(storer)
```
This will automatically update the cookie if the expiration has been changed by the user and this design has the nice side effect of being able to retrieve the expiration for each session with `storer.GetExpiration()`. As suggested, I put the expiration in the session provider DBs now.

Does that design look acceptable to you?

savsgio commented on mysql/dao.go:
```diff
@@ -97,17 +97,16 @@ func (db *Dao) deleteBySessionID(sessionID []byte) (int64, error) {
 // insert new session
-func (db *Dao) insert(sessionID, contents []byte, lastActiveTime int64) (int64, error) {
-	return db.Exec(db.sqlInsert, gotils.B2S(sessionID), gotils.B2S(contents), lastActiveTime)
+func (db *Dao) insert(sessionID, contents []byte, lastActiveTime int64, expiration int64) (int64, error) {
```
Use `time.Duration` instead of `int64`
Note: Propagate this change to postgres and sqlite3, please

clems4ever replied:
What about mysql?

clems4ever replied:
ack

clems4ever replied:
done

savsgio commented on mysql/dao.go:
```diff
@@ -97,17 +97,16 @@ func (db *Dao) deleteBySessionID(sessionID []byte) (int64, error) {
 
 // insert new session
-func (db *Dao) regenerate(oldID, newID []byte, lastActiveTime int64) (int64, error) {
+func (db *Dao) regenerate(oldID, newID []byte, lastActiveTime int64, expiration int64) (int64, error) {
```
Use time.Duration instead of int64
Note: Propagate this change to postgres and sqlite3, please

clems4ever replied:
ack

clems4ever replied:
done

savsgio commented on session.go:
```diff
@@ -151,21 +151,26 @@ func (s *Session) Get(ctx *fasthttp.RequestCtx) (Storer, error) {
 		return nil, errNotSetProvider
 	}
 
+	newSessionGenerated := false
```
This is unnecessary, revert it please.

clems4ever replied:
var newSessionGenerated bool instead?

savsgio replied:
I mean to not allocate a new variable to check if new session is generated. Put directly `s.setHTTPValues(ctx, sessionID)` inside condition, like this: https://github.com/fasthttp/session/blob/master/session.go#L161

clems4ever replied:
I need the expiration since the signature is now `setHTTPValues(ctx *fasthttp.RequestCtx, sessionID []byte, expires time.Duration)` and I need the store to get the expiration value. 

Otherwise I can revert the change and use the configured expiration from the global config but it's a weak design since now I think it's the responsibility of the store to decide whether it wants to consume the configured expiration or not.

I will propose something else to avoid the allocation but still get the expiration from the store.

clems4ever replied:
Well, actually the other solution was to declare the store before the if block and do an if/else and assign store in both blocks but this solution also pre-allocates the store variable anyway. I think the current solution is the best we can get from a architecture perspective but if you want I can avoid the allocation by setting the expiration coming from the config, it just creates a small weakness IMHO. I let you tell me.

savsgio replied:
Sorry, I don't see you get expiration from store.

clems4ever replied:
look up to line 172, you'll understand what I mean.

savsgio commented on memory/provider.go:
```diff
@@ -103,17 +103,14 @@ func (mp *Provider) Count() int {
 func (mp *Provider) GC() {
 	for _, kv := range mp.memoryDB.D {
-		if time.Now().Unix() >= (kv.Value.(*Store).lastActiveTime + mp.expiration) {
+		store := kv.Value.(*Store)
```
Check if expiration is `0` to non remove that session.

clems4ever replied:
done

clems4ever commented:
Great, I'm gonna fix all of your remarks.

Do you have any test setup with all providers to confirm everything is working fine? I've barely tested with the memory provider and I confirmed that sums in SQL filters like `last_active + expiration <= ?` work by executing a SQL query using a SQLite tool . But I'd like to be sure I've not broken anything. I can also manually test with redis as well today but I've no test setup for the rest.

clems4ever commented:
Fixes are done except one where I was not sure of what you meant.

savsgio commented:
> Great, I'm gonna fix all of your remarks.
> 
> Do you have any test setup with all providers to confirm everything is working fine? I've barely tested with the memory provider and I confirmed that sums in SQL filters like `last_active + expiration <= ?` work by executing a SQL query using a SQLite tool . But I'd like to be sure I've not broken anything. I can also manually test with redis as well today but I've no test setup for the rest.

Sorry, but now, i don't have any tests for that.

clems4ever commented:
I will do the tests with memory, Redis, SQLite and will try to spin up docker containers to test with other components. Hopefuly I'll be able to make another PR with a docker-compose so that everything spawns magically for anyone to at leat manually test with all providers.

savsgio commented on memcache/provider.go:
```diff
@@ -146,7 +151,10 @@ func (mcp *Provider) Regenerate(oldID, newID []byte) (session.Storer, error) {
 		newItem.Key = newKey
 		newItem.Value = oldItem.Value
-		newItem.Expiration = oldItem.Expiration
+		if mcp.expiration > math.MaxInt32 {
```
Check this in `Init()`, not here.

clems4ever replied:
you're right.

clems4ever replied:
done

savsgio commented on memcache/store.go:
```diff
@@ -11,7 +13,11 @@ func (mcs *Store) Save() error {
 	item.Value = value
-	item.Expiration = provider.expiration
+
+	if mcs.GetExpiration() > math.MaxInt32 {
```
Override `Store.SetExpiration`in this provider to check this.

savsgio replied:
Later, you could get expiration without checks.

clems4ever replied:
Can we safely consider this won't have any side effect regarding `store.HasExpirationChanged()`?
The expiration attribute is not exported so I'll need to call the parent `SetExpiration()` that can have side effects if called by someone else than the user.

savsgio replied:
Yes `SetExpiration()` is call by user. But this check is not necessary all time, only when user change the expiration.

Therefore, overwrite `Store.SetExpiration` to memcache provider in store.go, like `Store.Save()`

savsgio replied:
```go
func (mcs *Store) SetExpiration(expiration time.Duration) error {
    if expiration  > math.MaxInt32 {
        return errExpirationIsTooBig
    }
    mcs.newExpiration = expiration
    return nil
}
```

savsgio replied:
In  Store type add again newExpiration field

clems4ever replied:
Ok, I see. Two questions though: 
1/ SetExpiration does not return an error in the interface. Should I add it?
2/ I think I also need to override the GetExpiration and HasExpirationChanged. Do you confirm?

savsgio replied:
1) Yes
2) Why?

clems4ever replied:
1. OK
2. I'm a beginner in Go but if I'm not mistaken the attribute newExpiration in the new struct will not replace the original one from the parent struct, correct? If it's correct it means that GetExpiration and HasExpirationChanged should rely on the new attribute instead of the original from the parent struct. Where am I wrong? :)

savsgio replied:
Ohh yes, you need to re-implement it in memcache store. Because if you don't do it, when you get expiration, returns original field not new field in memcache

clems4ever replied:
Ok, great, we are in sync.

clems4ever replied:
But it leads to other issues then... I need to add two attributes: `expiration` and `newExpiration` because I need to compare them in HasExpirationChanged. This leads to also override the Init method... which in turns leads to also override all other methods... What do you think?

savsgio replied:
Oh no! Only re-add the expiration field like the newExpiration field, and overwrite methods in memcache store to get value from memcache store struct

savsgio replied:
You don't need to override Init of provider.

clems4ever replied:
@savsgio , I saw that you put some comments in my mailbox but I cannot see them in the discussion. Did you remove them in the meantime?

Do you see my point? I'm talking about the `Init` method of the Store not the Provider. It's because if we add two attributes to the inherited store (expiration and newExpiration), we need to be able to initialize them using `Init` of the Store as it's done for the parent store of other providers. Otherwise, only the parent attributes will be correctly initialized. Do you see what I mean?

savsgio replied:
Yes, i know.

Overwrite `Init` method too, because is the best solution i think.

clems4ever replied:
Done.

savsgio commented on memory/provider.go:
```diff
@@ -103,17 +103,19 @@ func (mp *Provider) Count() int {
-		if time.Now().Unix() >= (kv.Value.(*Store).lastActiveTime + mp.expiration) {
+		store := kv.Value.(*Store)
+		// Do not expire the session if expiration is set to 0
+		if store.GetExpiration() == 0 {
```
Save value in new variable, and use it, because expiration could be change by another gorutine.

clems4ever replied:
done

clems4ever commented:
I'm currently testing the providers, I've found two bugs already.

clems4ever commented:
I've tested all providers thanks to the docker-compose I added in the second commit. It made me find several bugs I introduced in the providers. The feature is now working well and doing what it should for: Memory, Postgres, MySQL, SQLite but Redis and Memcache are a bit trickier. 

The problem is that we cannot retrieve the expiration previously set in Memcache as you can read here: https://stackoverflow.com/questions/2558706/how-can-i-get-the-expire-time-for-the-particular-item-in-memcached. The expiration of the items is simply set to 0 when retrieved.
For Redis the problem is that the retrieved expiration is actually the number of seconds remaining before expiration of the key, so basically the original expiration duration is lost once set and so cannot be used in the next `Save` operation...

One way to workaround that would probably be to write the expiration along with the data in the value.
What do you think? How do you want me to proceed?

clems4ever commented:
I've tested the workaround I propose, it's working well and even simplifies the code for the memcache provider. I will create a branch in my fork for you to see what it's like.

savsgio commented on sqlite3/dao.go:
```diff
@@ -62,11 +62,13 @@ func (db *Dao) getSessionBySessionID(sessionID []byte) (*DBRow, error) {
 		return nil, err
 	}
 
+	data.expiration *= time.Second
```
it's not necessary the multiplication

savsgio commented on sqlite3/dao.go:
```diff
@@ -87,8 +89,8 @@ func (db *Dao) countSessions() int {
-func (db *Dao) updateBySessionID(sessionID, contents []byte, lastActiveTime int64) (int64, error) {
-	return db.Exec(db.sqlUpdateBySessionID, gotils.B2S(contents), lastActiveTime, gotils.B2S(sessionID))
+func (db *Dao) updateBySessionID(sessionID, contents []byte, lastActiveTime int64, expiration time.Duration) (int64, error) {
+	return db.Exec(db.sqlUpdateBySessionID, gotils.B2S(contents), lastActiveTime, expiration/time.Second, gotils.B2S(sessionID))
```
it's not necessary the division

savsgio commented on sqlite3/dao.go:
```diff
@@ -97,17 +99,16 @@ func (db *Dao) deleteBySessionID(sessionID []byte) (int64, error) {
-func (db *Dao) insert(sessionID, contents []byte, lastActiveTime int64) (int64, error) {
-	return db.Exec(db.sqlInsert, gotils.B2S(sessionID), gotils.B2S(contents), lastActiveTime)
+func (db *Dao) insert(sessionID, contents []byte, lastActiveTime int64, expiration time.Duration) (int64, error) {
+	return db.Exec(db.sqlInsert, gotils.B2S(sessionID), gotils.B2S(contents), lastActiveTime, expiration/time.Second)
```
it's not necessary the division

savsgio commented on sqlite3/dao.go:
```diff
@@ -97,17 +99,16 @@ func (db *Dao) deleteBySessionID(sessionID []byte) (int64, error) {
-func (db *Dao) regenerate(oldID, newID []byte, lastActiveTime int64) (int64, error) {
-	return db.Exec(db.sqlRegenerate, gotils.B2S(newID), lastActiveTime, gotils.B2S(oldID))
+func (db *Dao) regenerate(oldID, newID []byte, lastActiveTime int64, expiration time.Duration) (int64, error) {
+	return db.Exec(db.sqlRegenerate, gotils.B2S(newID), lastActiveTime, expiration/time.Second, gotils.B2S(oldID))
```
it's not necessary the division

clems4ever commented on memcache/provider.go:
```diff
@@ -84,7 +86,10 @@ func (mcp *Provider) Init(expiration int64, cfg session.ProviderConfig) error {
 	mcp.db = memcache.New(mcp.config.ServerList...)
 	mcp.db.MaxIdleConns = mcp.config.MaxIdleConns
-	mcp.expiration = int32(expiration)
+	if expiration/time.Second > math.MaxInt32 {
```
ack

clems4ever replied:
Actually this check is still needed to prevent a too big expiration set in configuration.

clems4ever commented on memcache/provider.go:
```diff
@@ -146,7 +155,9 @@ func (mcp *Provider) Regenerate(oldID, newID []byte) (session.Storer, error) {
-		newItem.Expiration = oldItem.Expiration
+		// Expiration can be converted safely from time.Duration to in32 because
+		// mcp.expiration has been checked to be safely castable in Init method.
+		newItem.Expiration = int32(mcp.expiration / time.Second)
```
it's necessary since the value stored is the default value which is 7200000000 and that does not fit in int32... Same issue for all providers, SQL providers cannot store such a value. It throws when running the example.

clems4ever replied:
So we either need to consider not using Duration anymore but an int32 representing the number of seconds or do those divisions. You choose.

savsgio replied:
mcp.expiration previously check in Provider.Init o int set.Expiration, here it's not neccesary to divide by seconds, because the value already be correct

clems4ever replied:
the value is in microseconds and cannot be stored as an int in DB or cache backend, it's too big.

clems4ever replied:
look at what `2 * time.Hour` produces. This is the default value for the expiration and it's expressed in microsecond and that overflows an int32. It's not even comparable with a Unix timestamp as last_activity is stored in DB.

clems4ever replied:
`DELETE FROM %s WHERE last_active+expiration<=$1` does not work if not converted in seconds. The driver throws because the value is too big. Try it out with the example, you'll see.

savsgio replied:
in SQL i agree, but in memcache you put a correct value on item, later the memcache library parse this value to the most convenient value for him

clems4ever replied:
Here for memcache the value should be expressed in seconds as well as per the documentation: https://github.com/golang/appengine/blob/master/memcache/memcache.go#L65

savsgio replied:
The type of memcache.Item.Expiration is a time.Duration

clems4ever replied:
you're right, but this code is newer... In you're current version it's an int32...

clems4ever replied:
So my guess is there was a bug hidden by the fact that the cookie expires before the data. A bump of the version of memcache would fix that :).

savsgio replied:
Oh of course, update the memcache library.

clems4ever replied:
well, the link I provided is not from the same library. In the one you're using it's actually an int32 even in master. https://github.com/bradfitz/gomemcache/blob/551aad21a6682b95329c1f5bd62ee5060d64f7e8/memcache/memcache.go#L167

savsgio replied:
Yeah, i see after, but don't worry. let's leave it with the current library

clems4ever commented on memcache/store.go:
```diff
@@ -11,11 +16,36 @@ func (mcs *Store) Save() error {
 	item.Key = provider.getMemCacheSessionKey(mcs.GetSessionID())
 	item.Value = value
-	item.Expiration = provider.expiration
+	item.Expiration = int32(mcs.GetExpiration() / time.Second)
```
explanation here: https://github.com/fasthttp/session/pull/9#discussion_r280490403

clems4ever commented on memcache/store.go:
```diff
@@ -11,11 +16,36 @@ func (mcs *Store) Save() error {
 	return err
 }
+
+// Init initialize the store
```
If I call SetExpiration from the parent, I make the method HasExpirationChanged() return true, which I don't want. That's why we needed to override those methods from the beginning. Anyway with the lastest info, I did another fix that will simplify this code, I will show you the branch and you'll tell me if what I have done suits you.

savsgio replied:
No because you always interact with parent's fields, this is only a wrapper to check max value of int32

clems4ever replied:
I agree. I will do that change.

clems4ever replied:
done

clems4ever commented on memcache/types.go:
```diff
@@ -34,12 +35,14 @@ type Config struct {
 type Store struct {
 	session.Store
+
+	newExpiration time.Duration
```
yep.

clems4ever replied:
done

savsgio commented on docker-compose.yml:
```diff
@@ -0,0 +1,35 @@
+version: '3'
```
Remove this file, please, and add later if it's necessary in other PR. 

Note: I would like to do unittest without bbdd services.

clems4ever replied:
done

savsgio commented:
Do you finish the PR?

clems4ever commented:
I made a last patch to fix an issue with Redis and Memcache session that was not correctly set. Now everything works as expected AFAIC.

savsgio commented on store.go:
```diff
@@ -1,8 +1,13 @@
 
+import "time"
+
+const expirationAttributeKey = "__fasthttp_session_expiration__"
```
Move this constant to `const.go`

clems4ever replied:
ack

savsgio commented on store.go:
```diff
@@ -71,6 +76,27 @@ func (s *Store) SetSessionID(id []byte) {
 
+// SetExpiration set expiration for the session
+func (s *Store) SetExpiration(expiration time.Duration) error {
+	s.expirationChanged = true
```
Use mutex for thread/gorutine-safe

clems4ever replied:
ack

savsgio commented on store.go:
```diff
@@ -71,6 +76,27 @@ func (s *Store) SetSessionID(id []byte) {
+
+// HasExpirationChanged check wether the expiration has been changed
+func (s *Store) HasExpirationChanged() bool {
+	return s.expirationChanged
```
Use mutex for thread/gorutine-safe

clems4ever replied:
ack

savsgio commented:
> I made a last patch to fix an issue with Redis and Memcache session that was not correctly set. Now everything works as expected AFAIC.

Good working! Now the code it's cleaner.
Please, fix the last comments and i will merge this PR.

clems4ever commented:
I'm done.

savsgio commented on types.go:
```diff
@@ -78,6 +78,7 @@ type Store struct {
 	data              *Dict
 	defaultExpiration time.Duration
 	expirationChanged bool
+	expirationLock    sync.RWMutex
```
Why do you duplicate this??
Please, use `lock`

clems4ever replied:
Alright, sorry.

clems4ever replied:
Done

clems4ever commented:
Do you want me to rebase?

savsgio commented:
> Do you want me to rebase?

Don't worry. Thank you.

clems4ever commented:
Yes! Thank you! It took some time but I think it's really worth it :).
